src/Glpi/Altcha/AltchaManager.php

```diff
@@ -0,0 +1,291 @@
+     * We must thus maintain a registry of solved challenges and reject any
+     * attempt to reuse a challenge that has already been accepted.
+     */
+    public function removeChallenge(string $raw_payload): void
+    {
+        if (!$this->isEnabled()) {
+            return;
+        }
+
+        try {
+            $payload = $this->decodePayload($raw_payload);
+            $challenge = $this->getChallengeHashFromPayload($payload);
+        } catch (InvalidPayloadException) {
+            throw new RuntimeException("No challenge found");
+        }
+
+        $this->removeOngoingChallenge($challenge);
+    }
+
+    /**
+     * Get the hmac key used to generate and validate challenge.
+     * The ideal key size depends on the hash algorithm used, by default
+     * the ChallengeOptions object specify the SHA-256 algorithm.
+     *
+     * SHA-256 has a 512 bits block size and a 256 bits output length.
+     * In general, we need a key smaller or equals to the block size and higher
+     * or equals to the output size (so <= 512 and >= 256 in our case).
+     *
+     * The existing glpicrypt.key will work fine for this as it is has a length
+     * of 256 bits at the time this comment was written.
+     */
+    private function computeHmacKey(): string
+    {
+        return (new GLPIKey())->get();
+    }
+
+    private function initSessionStorage(): void
+    {
+        if (!isset($_SESSION[self::SESSION_STORAGE_KEY])) {
+            $_SESSION[self::SESSION_STORAGE_KEY] = [];
+        }
+    }
+
+    private function storeOngoingChallenge(string $hash): void
+    {
+        $_SESSION[self::SESSION_STORAGE_KEY][$hash] = true;
+    }
+
+    private function hasOngoingChallenge(string $hash): bool
+    {
+        return $_SESSION[self::SESSION_STORAGE_KEY][$hash] ?? false;
+    }
+
+    private function removeOngoingChallenge(string $hash): void
+    {
+        unset($_SESSION[self::SESSION_STORAGE_KEY][$hash]);
+    }
+
+    /** @throws InvalidPayloadException */
+    private function decodePayload(string $payload): array
+    {
+        try {
+            $decoded = base64_decode($payload, true);
+            $data = json_decode($decoded, true, 2, JSON_THROW_ON_ERROR);
+        } catch (UrlException|JsonException) {
+            throw new InvalidPayloadException();
+        }
+
+        if (!is_array($data) || $data === []) {
+            throw new InvalidPayloadException();
+        }
+
+        return $data;
+    }
+
+    /** @throws InvalidPayloadException */
+    private function getChallengeHashFromPayload(array $payload): string
+    {
+        if (!isset($payload['challenge'])) {
+            throw new InvalidPayloadException();
+        }
+
+        return $payload['challenge'];
+    }
+
+    private function getMode(): AltchaMode
+    {
+        // @phpstan-ignore instanceof.alwaysTrue (don't trust custom user config)
+        if (!GLPI_ALTCHA_MODE instanceof AltchaMode) {
+            throw new RuntimeException();
+        }
```

Comment:
We should allow strings to ease the configuration.

```suggestion
        if (GLPI_ALTCHA_MODE instanceof AltchaMode) {
            $mode = GLPI_ALTCHA_MODE;
        } else {
            $mode = AltchaMode::tryFrom(GLPI_ALTCHA_MODE);
        }

        if (!$mode instanceof AltchaMode) {
            throw new RuntimeException();
        }
```